BSP/sharpMemoryLCD/LS013B7DH03.c

```diff
@@ -0,0 +1,206 @@
+    }
+
+    TIM_OC_InitTypeDef sConfig;
+    sConfig.OCMode       = TIM_OCMODE_PWM1;
+    sConfig.OCPolarity   = TIM_OCPOLARITY_HIGH;
+    sConfig.OCFastMode   = TIM_OCFAST_DISABLE;
+    sConfig.OCNPolarity  = TIM_OCNPOLARITY_HIGH;
+    sConfig.OCNIdleState = TIM_OCNIDLESTATE_RESET;
+    sConfig.OCIdleState  = TIM_OCIDLESTATE_RESET;
+    sConfig.Pulse = 1;
+    if (HAL_TIM_PWM_ConfigChannel(&TimHandle, &sConfig, TIM_CHANNEL_1) != HAL_OK)
+    {
+        assert_param(false);
+    }
+
+    if (HAL_TIM_PWM_Start(&TimHandle, TIM_CHANNEL_1) != HAL_OK)
+    {
+        assert_param(false);
+    }
+
+    SpiHandle.Instance               = SPI2;
+    SpiHandle.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_256; // ===> TODO set correct baudrate to 1Mbps
+    SpiHandle.Init.Direction         = SPI_DIRECTION_1LINE;
+    SpiHandle.Init.CLKPhase          = SPI_PHASE_1EDGE;
+    SpiHandle.Init.CLKPolarity       = SPI_POLARITY_LOW;
+    SpiHandle.Init.DataSize          = SPI_DATASIZE_8BIT;
+    SpiHandle.Init.FirstBit          = SPI_FIRSTBIT_LSB;
+    SpiHandle.Init.TIMode            = SPI_TIMODE_DISABLE;
+    SpiHandle.Init.CRCCalculation    = SPI_CRCCALCULATION_DISABLE;
+    SpiHandle.Init.NSS               = SPI_NSS_SOFT;
+    SpiHandle.Init.NSSPMode          = SPI_NSS_PULSE_DISABLE;
+    SpiHandle.Init.Mode              = SPI_MODE_MASTER;
+    if(HAL_SPI_Init(&SpiHandle) != HAL_OK)
+    {
+        assert_param(false);
+    }
+
+    sharpMemoryLCD_clearScreen();
+
+    LCDdisplayEnable(true);
+}
+
+
+void sharpMemoryLCD_clearScreen(void)
+{
+    /* To clear sreen send CMD_CLEAR (3bit) + 16 additional dummy bits */
+    uint8_t buffer[] = {CMD_CLEAR, 0x00};
+    LCDslaveSelect(true);
+    if(HAL_SPI_Transmit(&SpiHandle, buffer, sizeof(buffer), 1000) != HAL_OK)
+    {
+        assert_param(false);
+    }
+    LCDslaveSelect(false);
+}
+
+
+void sharpMemoryLCD_printTextLine(uint8_t line, const char *text, uint8_t nbChar)
+{
+    uint8_t buffer[Font16.Height*SCREEN_LENGTH/Font16.Width];
+    // this init could be optimized later by only init the end of line + using SIMD
+    for(uint16_t i = 0; i<(Font16.Height*SCREEN_LENGTH/Font16.Width); i++)
+    {
+        buffer[i] = 0;
+    }
+
+    if(line < (SCREEN_LENGTH/Font16.Width))
+    {
+        uint16_t sreenX=0;
+        uint16_t XbytesPerFontChar = (Font16.Width+7)/8;
+        uint16_t bytesPerFontChar = XbytesPerFontChar*Font16.Height;
+        //end display at first character out of ascii (most common will be \0) or out of screen
+        for(uint8_t c=0; (c<nbChar) && (' '<=text[c]) && (text[c]<='~') && (c<(SCREEN_LENGTH/Font16.Width)); c++)
+        {
+            // here SIMD instructions could be used to process 4 lines in one instruction
+            for(uint8_t x=0; x<Font16.Width; x++)
+            {
+                for(uint8_t y=0; y<Font16.Height; y++)
+                {
+                    uint16_t currentByteInFont = bytesPerFontChar*(text[c]-' ') + XbytesPerFontChar*y + x/8;
```

Comment:
This is pure dark magic to me 😄 
What about introducing a function?

Reply:
now, only god knows...

Reply:
added comments

Reply:
So this was an index all along. `indexByteInFont`